### PR TITLE
Target C++11

### PR DESCRIPTION
**Pull request by monatis**
Safely ignore changes in `image-search`, they are due to formatting other than a single line, which is simply the very same change in `clip.cpp`.

I'll also try using our own recursive directory walk function from `common-clip` instead of `std::filesystem`.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -141,7 +141,7 @@ if (CLIP_ALL_WARNINGS)
 
 endif()
 
-add_compile_options(-Wno-format -O3)
+add_compile_options(-Wno-format)
```
Comment (Green-Sky) [inline CMakeLists.txt:144]: move into the conditional above, this should fail for msvc

**Comment by monatis** [comment]
We don't need `std::filesystem` and C++17 anymore. It's also possible to search by image `./bin/image-search <search string or /path/to/query/image>`. If it's an image file we load the image from that path and encode it, otherwise we assume that it's regular text.

File: examples/common-clip.cpp [hunk 2]
```diff
@@ -114,6 +118,41 @@ std::map<std::string, std::vector<std::string>> get_dir_keyed_files(const std::s
     return result;
 }
 
+bool is_image_file_extension(std::string &path)
```
Comment (Green-Sky) [inline examples/common-clip.cpp:121]: make parameter const ref

File: examples/image-search/build.cpp [hunk 3]
```diff
@@ -103,49 +107,52 @@ int main(int argc, char** argv) {
...
         fprintf(stdout, "%s: starting base dir scan of '%s'\n", __func__, base_dir.c_str());
-
-        for (auto const& dir_entry : std::filesystem::recursive_directory_iterator(base_dir)) {
-            if (!dir_entry.is_regular_file()) {
-                continue;
+        auto results = get_dir_keyed_files(base_dir, 0);
+
+        for (auto &entry : results)
+        {
+            printf("%s: processing %d files in %s\n", __func__, entry.second.size(), entry.first.c_str());
```
Comment (Green-Sky) [inline examples/image-search/build.cpp:117]:
```
main: starting base dir scan of '../tests/'
main: processing 2 files in
..
```
why does it say "processing x files in" and then only dots? code looks fine o.o

Comment (Green-Sky) [reply examples/image-search/build.cpp:117]: verbose 2
```
main: starting base dir scan of '../tests/'
main: processing 2 files in
main: found image file '../tests//white.jpg'
main: found image file '../tests//red_apple.jpg'
```

File: examples/image-search/search.cpp [hunk 4]
```diff
@@ -27,59 +30,89 @@ void my_print_help(int argc, char **argv, my_app_params &params) {
...
-                params.search_text += " ";
-                params.search_text += argv[i];
+            if (i == argc - 1 && is_image_file_extension(arg))
+            {
+                params.img_path = arg;
+            }
+            else
+            {
+                params.search_text = arg;
+                for (++i; i < argc; i++)
```
Comment (Green-Sky) [inline examples/image-search/search.cpp:99]: you kept the ++i i++ :rofl:

**Review by Green-Sky — COMMENTED** [review]
I totally forgot about searching by image :+1:

File: examples/image-search/build.cpp (hunk 3 above)
Comment (monatis) [reply examples/image-search/build.cpp:117]: I just found printing every image file a little bit too verbose with default verbosity = 1 and used dots as an activity indicator, but maybe I can come up with something smarter.

File: examples/image-search/search.cpp (hunk 4 above)
Comment (monatis) [reply examples/image-search/search.cpp:99]: :D even the compiler should be surprised by it

File: examples/image-search/build.cpp (hunk 3 above)
Comment (Green-Sky) [reply examples/image-search/build.cpp:117]: no that is fine. the problem is the missing string after `files in` `processing %d files in %s\n`

**Pull request merged**